File: src/main/java/am/petstore/PetStore/user/entity/UserModel.java
package am.petstore.PetStore.user.entity;

import am.petstore.PetStore.order.entity.OrderEntity;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import am.petstore.PetStore.user.annotations.CheckEmail;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.util.Collection;
import java.util.Date;
import java.util.Set;

@Entity
@Table(name = "users")
@EntityListeners(AuditingEntityListener.class)
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class UserModel implements Serializable, UserDetails {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "password")
    @Size(min = 6)
    private String password;

    @Column(name = "email", unique = true)
    @CheckEmail
    private String email;

    @Column(name = "fullname")
    private String fullname;

    @Column(name = "username")
    @JsonIgnore
    private String username;

    @Column(name = "birthday")
    private String birthday;

    @Column(name = "photo")
    private String photo;

    @Column(name = "gender")
    private String gender;

    @Column(name = "active")
    @JsonIgnore
    private Boolean active;

    @Column(name = "firebase_id")
    private String firebase_id;

    @Column(name = "phone", unique = true)
    @Pattern(regexp = "[+0-9]*")
    @NotNull
    private String phone;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "created_at", nullable = false, updatable = false)
    @CreatedDate
    private Date createdAt;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "updated_at", nullable = false)
    @LastModifiedDate
    private Date updatedAt;

    @ElementCollection(targetClass = Role.class, fetch = FetchType.EAGER)
    @CollectionTable(name = "user_role", joinColumns = @JoinColumn(name = "user_id"))
    @Enumerated(EnumType.STRING)
    @JsonIgnore
    private Set<Role> roles;

    @OneToMany
    @JoinColumn(name = "user_id")
    private Set<Device> devices;

    @OneToMany
    @JoinColumn(name = "user_id")
    private Set<OrderEntity> orders;

    public UserModel() {
    }

    public UserModel(@Size(min = 6) String password, String email, String fullname, String birthday, String gender, Boolean active, String firebase_id, @Pattern(regexp = "[+0-9]*") @NotNull String phone, Date createdAt, Date updatedAt, Set<Role> roles, Set<Device> devices, Set<OrderEntity> orders) {
        this.password = password;
        this.email = email;
        this.fullname = fullname;
        this.birthday = birthday;
        this.gender = gender;
        this.active = active;
        this.firebase_id = firebase_id;
        this.phone = phone;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.roles = roles;
        this.devices = devices;
        this.orders = orders;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    @Override
    @JsonIgnore
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return roles;
    }

    public String getPassword() {
        return password;
    }

    @Override
    @JsonIgnore
    public String getUsername() {
        return phone;
    }

    @Override
    @JsonIgnore
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    @JsonIgnore
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    @JsonIgnore
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    @JsonIgnore
    public boolean isEnabled() {
        return active;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getFullname() {
        return fullname;
    }

    public void setFullname(String fullname) {
        this.fullname = fullname;
    }

    public String getBirthday() {
        return birthday;
    }

    public void setBirthday(String birthday) {
        this.birthday = birthday;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getFirebase_id() {
        return firebase_id;
    }

    public void setFirebase_id(String firebase_id) {
        this.firebase_id = firebase_id;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Date createdAt) {
        this.createdAt = createdAt;
    }

    public Date getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Date updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Set<Role> getRoles() {
        return roles;
    }

    public void setRoles(Set<Role> roles) {
        this.roles = roles;
    }

    public Set<Device> getDevices() {
        return devices;
    }

    public void setDevices(Set<Device> devices) {
        this.devices = devices;
    }

    @Override
    public String toString() {
        return "UserModel{" +
                "id=" + id +
                ", password='" + password + '\'' +
                ", email='" + email + '\'' +
                ", fullname='" + fullname + '\'' +
                ", birthday='" + birthday + '\'' +
                ", gender='" + gender + '\'' +
                ", firebase_id='" + firebase_id + '\'' +
                ", phone='" + phone + '\'' +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                ", roles=" + roles +
                ", devices=" + devices +
                ", orders=" + orders +
                '}';
    }

    public Set<OrderEntity> getOrders() {
        return orders;
    }

    public void setOrders(Set<OrderEntity> orders) {
        this.orders = orders;
    }
}

File: src/main/java/am/petstore/PetStore/user/entity/UserModel.kt
package am.petstore.PetStore.user.entity

import am.petstore.PetStore.order.entity.OrderEntity
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.io.Serializable
import java.util.*
import javax.persistence.*
import javax.validation.constraints.NotNull
import javax.validation.constraints.Pattern
import javax.validation.constraints.Size

@Entity
@Table(name = "users")
@EntityListeners(AuditingEntityListener::class)
@JsonIgnoreProperties("hibernateLazyInitializer", "handler")
class UserModel : Serializable, UserDetails {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
    @Column(name = "password")
    private var password: @Size(min = 6) String? = null
    @Column(name = "email", unique = true)
    var email: String? = null
    @Column(name = "fullname")
    var fullname: String? = null
    @Column(name = "username")
    @JsonIgnore
    private val username: String? = null
    @Column(name = "birthday")
    var birthday: String? = null
    @Column(name = "photo")
    private val photo: String? = null
    @Column(name = "gender")
    var gender: String? = null
    @Column(name = "active")
    @JsonIgnore
    private var active: Boolean? = null
    @Column(name = "firebase_id")
    var firebase_id: String? = null
    @Column(name = "phone", unique = true)
    var phone: @Pattern(regexp = "[+0-9]*") @NotNull String? = null
    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "created_at", nullable = false, updatable = false)
    @CreatedDate
    var createdAt: Date? = null
    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "updated_at", nullable = false)
    @LastModifiedDate
    var updatedAt: Date? = null
    @ElementCollection(targetClass = Role::class, fetch = FetchType.EAGER)
    @CollectionTable(name = "user_role", joinColumns = [JoinColumn(name = "user_id")])
    @Enumerated(EnumType.STRING)
    @JsonIgnore
    var roles: Set<Role>? = null
    @OneToMany
    @JoinColumn(name = "user_id")
    var devices: MutableSet<Device?>? = null
    @OneToMany
    @JoinColumn(name = "user_id")
    var orders: Set<OrderEntity>? = null

    constructor() {}
    constructor(password: @Size(min = 6) String?, email: String?, fullname: String?, birthday: String?, gender: String?, active: Boolean?, firebase_id: String?, phone: @Pattern(regexp = "[+0-9]*") @NotNull String?, createdAt: Date?, updatedAt: Date?, roles: Set<Role>?, devices: MutableSet<Device?>?, orders: Set<OrderEntity>?) {
        this.password = password
        this.email = email
        this.fullname = fullname
        this.birthday = birthday
        this.gender = gender
        this.active = active
        this.firebase_id = firebase_id
        this.phone = phone
        this.createdAt = createdAt
        this.updatedAt = updatedAt
        this.roles = roles
        this.devices = devices
        this.orders = orders
    }

    @JsonIgnore
    override fun getAuthorities(): Collection<GrantedAuthority> {
        return roles!!
    }

    override fun getPassword(): String {
        return password!!
    }

    @JsonIgnore
    override fun getUsername(): String {
        return phone!!
    }

    @JsonIgnore
    override fun isAccountNonExpired(): Boolean {
        return true
    }

    @JsonIgnore
    override fun isAccountNonLocked(): Boolean {
        return true
    }

    @JsonIgnore
    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    @JsonIgnore
    override fun isEnabled(): Boolean {
        return active!!
    }

    fun setPassword(password: String?) {
        this.password = password
    }

    override fun toString(): String {
        return "UserModel{" +
                "id=" + id +
                ", password='" + password + '\'' +
                ", email='" + email + '\'' +
                ", fullname='" + fullname + '\'' +
                ", birthday='" + birthday + '\'' +
                ", gender='" + gender + '\'' +
                ", firebase_id='" + firebase_id + '\'' +
                ", phone='" + phone + '\'' +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                ", roles=" + roles +
                ", devices=" + devices +
                ", orders=" + orders +
                '}'
    }

}